### ISSUE TITLE
WARNING:insights.tests:Replacing ps_auxcww in --appdebug output

### ISSUE THREAD

**Issue by PaulWay** [RedHatInsights/insights-core#1007]
Running the test suite on the Insights Plugins Demo ruleset with the options '-s -v --appdebug' gives multiple warnings in the output of the form:

```
WARNING:insights.tests:Replacing ps_auxcww
WARNING:insights.tests:['  File "/home/pwayper/Code/insights/bin/py.test", line 9, in <module>\n    load_entry_point(\'pytest==3.2.5\', \'console_scripts\', \'py.test\')()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/config.py", line 58, in main\n    return config.hook.pytest_cmdline_main(config=config)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__\n    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec\n    return self._inner_hookexec(hook, methods, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>\n    _MultiCall(methods, kwargs, hook.spec_opts).execute()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute\n    res = hook_impl.function(*args)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 140, in pytest_cmdline_main\n    return wrap_session(config, _main)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 111, in wrap_session\n    session.exitstatus = doit(config, session) or 0\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 146, in _main\n    config.hook.pytest_collection(session=session)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__\n    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec\n    return self._inner_hookexec(hook, methods, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>\n    _MultiCall(methods, kwargs, hook.spec_opts).execute()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute\n    res = hook_impl.function(*args)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 156, in pytest_collection\n    return session.perform_collect()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 654, in perform_collect\n    items = self._perform_collect(args, genitems)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 691, in _perform_collect\n    self.items.extend(self.genitems(node))\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/main.py", line 830, in genitems\n    rep = collect_one_node(node)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/runner.py", line 504, in collect_one_node\n    rep = ihook.pytest_make_collect_report(collector=collector)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__\n    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec\n    return self._inner_hookexec(hook, methods, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>\n    _MultiCall(methods, kwargs, hook.spec_opts).execute()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute\n    return _wrapped_call(hook_impl.function(*args), self.execute)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 248, in _wrapped_call\n    call_outcome = _CallOutcome(func)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__\n    self.result = func()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute\n    res = hook_impl.function(*args)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/runner.py", line 371, in pytest_make_collect_report\n    \'collect\')\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/runner.py", line 191, in __init__\n    self.result = func()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/runner.py", line 370, in <lambda>\n    lambda: list(collector.collect()),\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/python.py", line 389, in collect\n    return super(Module, self).collect()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/python.py", line 330, in collect\n    res = self.makeitem(name, obj)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/python.py", line 342, in makeitem\n    collector=self, name=name, obj=obj)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__\n    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec\n    return self._inner_hookexec(hook, methods, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>\n    _MultiCall(methods, kwargs, hook.spec_opts).execute()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute\n    return _wrapped_call(hook_impl.function(*args), self.execute)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call\n    wrap_controller.send(call_outcome)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/python.py", line 188, in pytest_pycollect_makeitem\n    res = list(collector._genfunctions(name, obj))\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/python.py", line 360, in _genfunctions\n    dict(metafunc=metafunc))\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 761, in call_extra\n    return self(**kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__\n    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec\n    return self._inner_hookexec(hook, methods, kwargs)\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>\n    _MultiCall(methods, kwargs, hook.spec_opts).execute()\n', '  File "/home/pwayper/Code/insights/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute\n    res = hook_impl.function(*args)\n', '  File "/home/pwayper/Code/insights/insights-plugins-demo/demo/tests/integration.py", line 12, in pytest_generate_tests\n    integration.generate_tests(metafunc, test_integration, "demo.tests", pattern=pattern)\n', '  File "/home/pwayper/Code/insights/insights-core/insights/tests/integration.py", line 37, in generate_tests\n    for t in ts:\n', '  File "/home/pwayper/Code/insights/insights-core/insights/tests/__init__.py", line 278, in __wrap\n    for input_data, expected in itertools.islice(func(), None, None, stride):\n', '  File "/home/pwayper/Code/insights/insights-plugins-demo/demo/tests/test_browser_root.py", line 56, in integration_tests\n    yield InputData(i).add(Specs.ps_auxcww, input_value), output_value\n', '  File "/home/pwayper/Code/insights/insights-core/insights/tests/__init__.py", line 185, in add\n    logger.warn(traceback.format_stack())\n']
```

Unpacking this gives a 96-line backtrace, the salient parts of which are:

```
File "/home/pwayper/Code/insights/insights-plugins-demo/demo/tests/integration.py", line 12, in pytest_generate_tests
    integration.generate_tests(metafunc, test_integration, "demo.tests", pattern=pattern)
File "/home/pwayper/Code/insights/insights-core/insights/tests/integration.py", line 37, in generate_tests
    for t in ts:
File "/home/pwayper/Code/insights/insights-core/insights/tests/__init__.py", line 278, in __wrap
    for input_data, expected in itertools.islice(func(), None, None, stride):
File "/home/pwayper/Code/insights/insights-plugins-demo/demo/tests/test_browser_root.py", line 56, in integration_tests
    yield InputData(i).add(Specs.ps_auxcww, input_value), output_value
File "/home/pwayper/Code/insights/insights-core/insights/tests/__init__.py", line 185, in add
    logger.warn(traceback.format_stack())
```

This warning is generated when the `InputData` object if `dr.get_delegate(spec).multi_output` is not set - regardless of whether the spec is already in `self.data`.

I suspect this code needs to be something like:
```python
        if dr.get_delegate(spec).multi_output:
            if spec not in self.data:
                self.data[spec] = []
            self.data[spec].append(content_provider)
        else:
            if spec in self.data:
                logger.warn("Replacing %s", spec.__name__)
                logger.warn(traceback.format_stack())
            self.data[spec] = content_provider
```

**Comment by csams** [comment]
Good catch! Fixed in 767a0a6e2d6089ef4a1bc13a5a6345268430fbf0.

**Comment by PaulWay** [comment]
Thanks @csams :-)
